core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
+def get_public_key_hash(keychain: seed.Keychain, path: List[int]) -> bytes:
+    node = keychain.derive(path)
+    public_key = remove_ed25519_prefix(node.public_key())
+    return hashlib.blake2b(data=public_key, outlen=28).digest()
 
 
-def is_safe_output_address(address) -> bool:
-    """
-    Determines whether it is safe to include the address as-is as
-    a tx output, preventing unintended side effects (e.g. CBOR injection)
-    """
-    address_data_encoded = _get_address_data(address)
-    if address_data_encoded is None:
-        return False
+def path_to_staking_path(path: List[int]) -> List[int]:
+    return path[:3] + [2, 0]
 
-    return _encode_address_raw(address_data_encoded) == address
 
+def derive_address(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+    human_readable: bool = True,
+) -> str:
+    is_byron_address = parameters.address_type == CardanoAddressType.BYRON_ADDRESS
 
-def _get_address_data(address: str) -> bytes:
-    try:
-        address_hex = base58.decode(address)
-        address_unpacked = cbor.decode(address_hex)
-    except ValueError as e:
-        if __debug__:
-            log.exception(__name__, e)
-        return None
+    if is_byron_address:
+        address = _derive_byron_address(keychain, parameters.path, protocol_magic)
+    else:
+        address = _derive_shelley_address(keychain, parameters, network_id)
 
-    if not isinstance(address_unpacked, list) or len(address_unpacked) != 2:
-        return None
+    if human_readable:
+        return get_human_readable_address(address, is_byron_address)
+    else:
+        return address
 
-    address_data_encoded = address_unpacked[0]
-    if not isinstance(address_data_encoded, bytes):
-        return None
 
-    return address_data_encoded
+def _derive_byron_address(
+    keychain: seed.Keychain, path: List[int], protocol_magic: int
+) -> bytes:
+    if not _is_byron_address_path(path):
+        raise wire.DataError("Invalid path for byron address!")
 
+    address = derive_byron_address(keychain, path, protocol_magic)
+    return address
 
-def matches_with_protocol_magic(address: str, protocol_magic: int) -> bool:
-    address_data_encoded = _get_address_data(address)
-    if address_data_encoded is None:
-        return False
 
-    address_data = cbor.decode(address_data_encoded)
-    if not isinstance(address_data, list) or len(address_data) < 2:
-        return False
+def _is_byron_address_path(path: List[int]) -> bool:
+    return path[0] == BYRON_PURPOSE
 
-    address_attributes = address_data[1]
 
-    if len(address_attributes) == 0:
-        return protocol_magic == ProtocolMagics.MAINNET
+def _derive_shelley_address(
+    keychain: seed.Keychain, parameters: CardanoAddressParametersType, network_id: int,
+) -> bytes:
+    if not _is_shelley_address_path(parameters.path):
+        raise wire.DataError("Invalid path for shelley address!")
```

Comment:
Resolved offline - I don't like the idea of having duplicated logic because of early validation. So we'll keep it like this and we'll see if Trezor guys share the same concerns.